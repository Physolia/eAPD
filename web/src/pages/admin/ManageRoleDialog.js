import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { Dialog, Dropdown, Button } from '@cmsgov/design-system';

const ManageRoleDialog = ({
  roleTypes,
  roleDescription,
  selectedAffiliation,
  hideManageModal,
  handleAffiliationUpdate
}) => {
  const findRoleByName = name => {
    return roleTypes.find(element => {
      return element.name === name;
    });
  };

  const getRole = selectedRoleName => {
    if (selectedRoleName) {
      return findRoleByName(selectedRoleName) || roleTypes[0];
    }
    return roleTypes[0];
  };

  const getRoleName = selectedRoleName => {
    const role = getRole(selectedRoleName);
    return role ? role.name : '';
  };

  const getRoleId = selectedRoleName => {
    const role = getRole(selectedRoleName);
    return role ? role.id : 0;
  };

  const [roleSelectedName, setRoleSelectedName] = useState(
    getRoleName(selectedAffiliation.role)
  );
  const [roleSelectedId, setRoleSelectedId] = useState(
    getRoleId(selectedAffiliation.role)
  );

  const handleDropdownUpdate = event => {
    setRoleSelectedName(event.target.value);
    const role = findRoleByName(event.target.value);
    setRoleSelectedId(role.id);
  };

  const handleUpdate = () => {
    // Dialog gets the selected affiliation
    // then on save it will send the role by id
    handleAffiliationUpdate(roleSelectedId);
  };

  const dropdownOptions = roleTypes.map(role => {
    const updatedRole = {
      ...role
    };
    updatedRole.key = role.name;
    updatedRole.value = role.name;
    updatedRole.label = role.name;
    return updatedRole;
  });

  dropdownOptions.unshift({
    label: 'Select an option',
    value: '',
    key: 'selectAnOption'
  });

  return (
    <Dialog
      onExit={hideManageModal}
      heading="Edit Role"
      actions={[
        <Button
          className="ds-u-margin-right--3 ds-u-margin-top--2"
          onClick={handleUpdate}
          key="action1"
        >
          Save
        </Button>,
        <Button variation="danger" onClick={hideManageModal} key="action2">
          Cancel
        </Button>
      ]}
    >
      <p>
        <strong>Name</strong> {selectedAffiliation.displayName}
      </p>
      <p>
        <strong>Phone Number</strong> {selectedAffiliation.primaryPhone}
      </p>
      <p>
        <strong>Email</strong> {selectedAffiliation.email}
      </p>
      <Dropdown
        options={dropdownOptions}
        hint={roleDescription}
        size="medium"
        label="Role"
        name="selectedPermission"
        value={roleSelectedName}
        onChange={handleDropdownUpdate}
      />
    </Dialog>
  );
};

ManageRoleDialog.propTypes = {
  roleTypes: PropTypes.array.isRequired,
  roleDescription: PropTypes.string,
  selectedAffiliation: PropTypes.object.isRequired,
  hideManageModal: PropTypes.func.isRequired,
  handleAffiliationUpdate: PropTypes.func.isRequired
};

ManageRoleDialog.defaultProps = {
  roleDescription: ''
};

export default ManageRoleDialog;
