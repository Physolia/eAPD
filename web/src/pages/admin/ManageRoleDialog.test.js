import React from 'react';
import { shallow } from 'enzyme';
import ManageRoleDialog from './ManageRoleDialog';

const selectedAffiliation = {
  displayName: 'Liz Lemon',
  email: 'liz@lemon.com',
  id: 24,
  mobilePhone: null,
  primaryPhone: '4045555555',
  role: null,
  secondEmail: null,
  stateId: 'md',
  status: 'requested',
  userId: '00u5mfj967KsdvBBB297'
};

describe('<ManageRoleDialog />', () => {
  const props = {
    roleTypes: [{ name: 'State Admin' }, { name: 'State Contractor' }],
    selectedAffiliation,
    hideManageModal: jest.fn(),
    handleAffiliationUpdate: jest.fn()
  };

  test('matches snapshot', () => {
    expect(shallow(<ManageRoleDialog {...props} />)).toMatchSnapshot();
  });

  test('renders dialog box title', () => {
    const component = shallow(<ManageRoleDialog {...props} />);
    expect(component.find('Edit Role')).toBeTruthy();
  });

  test('renders affiliation information', () => {
    const component = shallow(<ManageRoleDialog {...props} />);
    expect(component.find('p').at(0).text()).toEqual(
      `Name ${selectedAffiliation.displayName}`
    );
    expect(component.find('p').at(1).text()).toEqual(
      `Phone Number ${selectedAffiliation.primaryPhone}`
    );
    expect(component.find('p').at(2).text()).toEqual(
      `Email ${selectedAffiliation.email}`
    );
  });

  test('renders save and cancel buttons', () => {
    const component = shallow(<ManageRoleDialog {...props} />);
    expect(component.find('Save')).toBeTruthy();
    expect(component.find('Cancel')).toBeTruthy();
  });
});
