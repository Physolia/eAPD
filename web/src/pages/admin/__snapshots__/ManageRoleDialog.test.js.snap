// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageRoleDialog /> matches snapshot 1`] = `
<Dialog
  actions={
    Array [
      <Button
        className="ds-u-margin-right--3 ds-u-margin-top--2"
        component="button"
        onClick={[Function]}
        type="button"
      >
        Save
      </Button>,
      <Button
        component="button"
        onClick={[MockFunction]}
        type="button"
        variation="danger"
      >
        Cancel
      </Button>,
    ]
  }
  ariaCloseLabel="Close modal dialog"
  closeButtonText="Close"
  closeButtonVariation="transparent"
  closeIcon={<CloseIcon />}
  escapeExitDisabled={false}
  escapeExits={true}
  heading="Edit Role"
  onExit={[MockFunction]}
  underlayClickExits={false}
>
  <p>
    <strong>
      Name
    </strong>
     
    Liz Lemon
  </p>
  <p>
    <strong>
      Phone Number
    </strong>
     
    4045555555
  </p>
  <p>
    <strong>
      Email
    </strong>
     
    liz@lemon.com
  </p>
  <Dropdown
    hint=""
    label="Role"
    name="selectedPermission"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "selectAnOption",
          "label": "Select an option",
          "value": "",
        },
        Object {
          "key": "State Admin",
          "label": "State Admin",
          "name": "State Admin",
          "value": "State Admin",
        },
        Object {
          "key": "State Contractor",
          "label": "State Contractor",
          "name": "State Contractor",
          "value": "State Contractor",
        },
      ]
    }
    size="medium"
    value="State Admin"
  />
</Dialog>
`;
